File: awsomepipeline/pipeline_stack.py
from aws_cdk import (
    aws_codepipeline as codepipeline,
    aws_codebuild as codebuild,
    aws_codepipeline_actions as codepipeline_actions,
    aws_kms as kms,
    aws_s3 as s3,
    aws_iam as iam,
    core
)


class PipelineStack(core.Stack):

    def __init__(self, scope: core.Construct, id: str, git_token_key: str, github_owner: str, github_repo: str,
                 github_branch: str, **kwargs) -> None:
        super().__init__(scope, id, **kwargs)

        role = iam.Role(
            self,
            "Role",
            assumed_by=iam.ServicePrincipal("codebuild.amazonaws.com")
        )

        role.add_managed_policy(
            iam.ManagedPolicy.from_aws_managed_policy_name("AdministratorAccess")
        )

        cdk_project = codebuild.PipelineProject(
            self,
            "Codebuild",
            build_spec=codebuild.BuildSpec.from_source_filename("codebuild/buildspec.yaml"),
            cache=codebuild.Cache.bucket(s3.Bucket(self, "Bucket")),
            environment=codebuild.BuildEnvironment(
                build_image=codebuild.LinuxBuildImage.STANDARD_2_0,
                privileged=True
                ),
            role=role
        )

        source_output = codepipeline.Artifact()
        staging_output = codepipeline.Artifact()
        production_output = codepipeline.Artifact()

        source_action = codepipeline_actions.GitHubSourceAction(
            action_name="GitHub_Source",
            owner=github_owner,
            repo=github_repo,
            branch=github_branch,
            oauth_token=core.SecretValue.secrets_manager(git_token_key),
            output=source_output
        )

        staging_action = codepipeline_actions.CodeBuildAction(
            action_name="Deliver",
            project=cdk_project,
            input=source_output,
            outputs=[staging_output],
            environment_variables={
                "ENV": {"value": "stg"}
            }
        )

        manual_approval_action = codepipeline_actions.ManualApprovalAction(
            action_name="Approve"
        )

        production_action = codepipeline_actions.CodeBuildAction(
            action_name="Deliver",
            project=cdk_project,
            input=source_output,
            outputs=[production_output],
            environment_variables={
                "ENV": {"value": "prd"}
            }
        )

        key = kms.Key(self, "key")
        bucket = s3.Bucket(self, "bucket_artifacts", encryption_key=key)
        pipeline = codepipeline.Pipeline(self, "Pipeline", artifact_bucket=bucket)
        pipeline.add_stage(stage_name="Source", actions=[source_action])
        pipeline.add_stage(stage_name="Staging", actions=[staging_action])
        pipeline.add_stage(stage_name="Approval", actions=[manual_approval_action])
        pipeline.add_stage(stage_name="Production", actions=[production_action])

File: infrastructure/lib/pipeline-stack.ts
import 'source-map-support/register'
import { Stack, StackProps, Duration, CfnOutput, SecretValue, RemovalPolicy } from 'aws-cdk-lib';
import { aws_iam, aws_codebuild, aws_codepipeline, aws_codepipeline_actions, aws_s3 } from 'aws-cdk-lib';
import { Construct } from 'constructs';

export class PipelineStack extends Stack {
    constructor(scope: Construct, id: string, gitToken: string, github_owner: string, github_repo: string,
        github_branch: string, props?: StackProps) {
        super(scope, id, props)

        const role = new aws_iam.Role(this, 'role', { assumedBy: new aws_iam.ServicePrincipal('codebuild.amazonaws.com') })
        role.addManagedPolicy(aws_iam.ManagedPolicy.fromAwsManagedPolicyName('AdministratorAccess'))

        const project = new aws_codebuild.PipelineProject(this, 'pipelineProject', {
            buildSpec: aws_codebuild.BuildSpec.fromSourceFilename('infrastructure/codebuild/buildspec.yaml'),
            cache: aws_codebuild.Cache.bucket(
                new aws_s3.Bucket(this, 'cache', {
                    removalPolicy: RemovalPolicy.DESTROY, 
                    autoDeleteObjects: true
                })
            ),
            environment: {
                buildImage: aws_codebuild.LinuxBuildImage.STANDARD_4_0,
                privileged: true
            },
            role: role
        })

        const source_output = new aws_codepipeline.Artifact()
        const staging_output = new aws_codepipeline.Artifact()
        const production_output = new aws_codepipeline.Artifact()

        const source_action = new aws_codepipeline_actions.GitHubSourceAction({
            actionName: 'GitHub_Source',
            owner: github_owner,
            repo: github_repo,
            branch: github_branch,
            oauthToken: SecretValue.secretsManager(gitToken),
            output: source_output
        })

        const staging_action = new aws_codepipeline_actions.CodeBuildAction({
            actionName: 'Deliver',
            project: project,
            input: source_output,
            outputs: [staging_output],
            environmentVariables: {
                'ENV': {
                    value: 'stg',
                },
                'WORKING_BRANCH': {
                    value: github_branch
                }
            }
        })

        const manual_approval_action = new aws_codepipeline_actions.ManualApprovalAction({
            actionName: 'Approve'
        })

        const production_action = new aws_codepipeline_actions.CodeBuildAction({
            actionName: 'Deliver',
            project: project,
            input: source_output,
            outputs: [production_output],
            environmentVariables: {
                'ENV': {
                    value: 'prd',
                },
                'WORKING_BRANCH': {
                    value: github_branch
                }
            }
        })

        const bucketArtifacts = new aws_s3.Bucket(this, 'artifacts', {
            removalPolicy: RemovalPolicy.DESTROY, 
            autoDeleteObjects: true
        })

        const pipeline = new aws_codepipeline.Pipeline(this, "Pipeline", {
            artifactBucket: bucketArtifacts
        })

        pipeline.addStage({
            stageName: 'Source',
            actions: [source_action]
        })

        pipeline.addStage({
            stageName: 'Staging',
            actions: [staging_action]
        })

        pipeline.addStage({
            stageName: 'Approval',
            actions: [manual_approval_action]
        })

        pipeline.addStage({
            stageName: 'Production',
            actions: [production_action]
        })

        new CfnOutput(this, 'LinkCodePipelinePage', {
            value: "https://"
                + this.region
                + ".console.aws.amazon.com/codesuite/"
                + "codepipeline/pipelines/"
                + pipeline.pipelineName
                + "/view?region=" + this.region
        })
    }
}
